src/main/kotlin/com/hj/leetcode/kotlin/problem1579/Solution.kt

```diff
@@ -0,0 +1,127 @@
+         * Create two UnionFind, one for Alice and another for Bob
+         */
+        val alice = UnionFind(n + 1) // additional size accounts for the unused label 0
+        val bob = UnionFind(n + 1) // additional size accounts for the unused label 0
+
+        // Use a variable to count the number of edges that can be removed
+        var numRemovedEdges = 0
+
+        // First, iterate all the type 3 edges, which can be traversed by both Alice and Bob
+        for ((type, u, v) in edges) {
+            if (type != 3) {
+                continue
+            }
+            // Try to union the two nodes of the edge for both Alice and Bob
+            val isRequiredEdge = alice.union(u, v)
+            bob.union(u, v)
+            /* If the union is no-op, the two nodes have already been connected, so the edge can be removed,
+             * and we increase the number of removed edge by 1
+             */
+            if (!isRequiredEdge) {
+                numRemovedEdges++
+            }
+        }
+
+        // Iterate all the type 1 edges, which can be traversed by Alice only
+        for ((type, u, v) in edges) {
+            if (type != 1) {
+                continue
+            }
+            // Try to union the two nodes of the edge for Alice
+            val isRequiredEdge = alice.union(u, v)
+            /* If the union is no-op, the two nodes have already been connected, so the edge can be removed,
+             * and we increase the number of removed edge by 1
+             */
+            if (!isRequiredEdge) {
+                numRemovedEdges++
+            }
+        }
+        /* Check the number of connected components in Alice. If it is not one, then Alice cannot fully
+         * traverse the graph and should return -1
+         */
+        val aliceNumComponents = alice.numComponents - 1 // Reduce 1 for the unused label 0
+        if (aliceNumComponents != 1) {
+            return -1
+        }
+
+        // Iterate all the type 2 edges, which can be traversed by Bob only
+        for ((type, u, v) in edges) {
+            if (type != 2) {
+                continue
+            }
+            // Try to union the two nodes of edge for Bob
+            val isRequiredEdge = bob.union(u, v)
+            /* If the union is no-op, the two nodes have already been connected, so the edge can be removed,
+             * and we increase the number of removed edge by 1
+             */
+            if (!isRequiredEdge) {
+                numRemovedEdges++
+            }
+        }
+        /* Check the number of connected components in Bob. If it is not one, then Bob cannot fully
+         * traverse the graph and should return -1
+         */
+        val bobNumComponents = bob.numComponents - 1 // Reduce 1 for the unused label 0
+        if (bobNumComponents != 1) {
+            return -1
+        }
+
+        // Return the number of edges that can be removed
+        return numRemovedEdges
+    }
```

Comment:
The main function is too long and should be decomposed into several sub-functions